### PR TITLE
Add start and stop callbacks

### PR DESCRIPTION
**Pull request by jevin**
This PR adds lifecycle events to Sinatra. Related issue: #1184

### CODE REVIEW

**Comment by jevin** [comment]
This is still missing tests. I'm not sure where they fit it, anyone can give me some pointers?

**Comment by jevin** [comment]
Okay this is ready for a review! @zzak (sorry I'm not sure who else to @')

**Comment by dentarg** [comment]
The code looks good to me.

This is only used if you use Sinatra to start you web server, is that correct?

Can you add some documentation to the README? Maybe even add an example of this to the `examples` directory?

Thanks

**Comment by jevin** [comment]
Hey @dentarg, sorry I missed your comment!

> This is only used if you use Sinatra to start you web server, is that correct?

Correct!

> Can you add some documentation to the README? Maybe even add an example of this to the `examples` directory?

I will get on that!

**Comment by jevin** [comment]
@dentarg I just updated the README and added `lifecycle_events.rb` in `examples`!

Also, if you feel like the lifecycle events would make sense for servers NOT started with Sinatra, I'm down to create another PR for this!

File: lib/sinatra/base.rb [hunk 1]
```diff
@@ -1559,6 +1567,8 @@ def quit!
         warn '== Sinatra has ended his set (crowd applauds)' unless suppress_messages?
         set :running_server, nil
         set :handler_name, nil
+
+        on_stop_callback.call if !on_stop_callback.nil?
```
Comment (dentarg) [inline lib/sinatra/base.rb:1571]:
```suggestion
        on_stop_callback.call unless on_stop_callback.nil?
```

File: lib/sinatra/base.rb [hunk 2]
```diff
@@ -1646,7 +1656,7 @@ def start_server(handler, server_settings, handler_name)
           set :running_server, server
           set :handler_name,   handler_name
           server.threaded = settings.threaded if server.respond_to? :threaded=
-
+          on_start_callback.call if !on_start_callback.nil?
```
Comment (dentarg) [inline lib/sinatra/base.rb:1659]:
```suggestion
          on_start_callback.call unless on_start_callback.nil?
```

**Review by dentarg — COMMENTED** [review]
I think this reads easier

File: test/server_test.rb [hunk 3]
```diff
@@ -27,6 +27,14 @@ def stop
   register 'mock', 'Rack::Handler::Mock'
 end
 
+class EventHookTest
+  def self.start
+  end
+
+  def self.stop
+  end
+end
```
Comment (dentarg) [inline test/server_test.rb:36]: I think you can make this an anonymous class only available inside `ServerTest`: https://dev.to/factorial/a-trick-with-ruby-anonymous-classes-11pp

Comment (jevin) [reply test/server_test.rb]: Sure thing! I've turned it into an anonymous class.

**Comment by jevin** [comment]
@dentarg, I've implemented the changes. Have a look and let me know!

**Comment by dentarg** [comment]
Great, do you mind rebasing to get CI green?

**Pull request merged**